packages/gatsby/src/datastore/in-memory/indexing.ts

```diff
@@ -21,6 +24,88 @@ export type FilterOp =  // TODO: merge with DbComparator ?
 // Note: `undefined` is an encoding for a property that does not exist
 
 export type FilterCacheKey = string
+type GatsbyNodeID = string
+
+export interface IGatsbyNodePartial {
+  isGatsbyNodePartial: boolean
+  id: GatsbyNodeID
+  internal: {
+    counter: number
+  }
+  indexFields: Set<string>
+  [k: string]: any
```

Comment:
what if `indexFields` values actually include `isGatsbyNodePartial` or `indexFields` ( the former lest likely, the latter I feel more likely to happen ) - should we maybe put those index values under nested object and not in the root of partial to avoid colissions?

Consider this case
```
const node = {
  id: 'some-id',
  isGatsbyNodePartial: true, // there is nothing preventing users/plugins using field name like that
  indexFields: ['In the node this is array because why not'],
  internal: {
    type: 'SomeType',
    contentDigest: 'some-content-digest',
    counter: 42
  }
}

const partial = getGatsbyNodePartial(node, [`indexFields`, `isGatsbyNodePartial`], {})
// how the partial looks like now? do we overwrite fields on node with our "metadata" or the other way around?
```